### PR TITLE
Add message passing to core

### PR DESCRIPTION
**Pull request by calumats**
- Prerequisite for PRO-486/PRO-619
 - Both NATS and Kinesis supported

### CODE REVIEW

File: common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  import io.circe.generic.semiauto._
+  import org.genivi.sota.marshalling.CirceInstances._
+
+  implicit val EncoderInstance: Encoder[DeviceSeenMessage] = deriveEncoder
+  implicit val DecoderInstance: Decoder[DeviceSeenMessage] = deriveDecoder
+}
+
+object MessageBusManager {
+
+  val log = LoggerFactory.getLogger(this.getClass)
```
Comment (koshelev) [inline common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala]: use akka logger

File: common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  def start(system: ActorSystem, config: Config): ConfigException Xor (DeviceSeenMessage => Unit) = {
+    def startNats(system: ActorSystem, config: Config): ConfigException Xor (DeviceSeenMessage => Unit) = {
```
Comment (koshelev) [inline common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala]: Why passing these parameters?

File: common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      log.trace("Starting NATS")
```
Comment (koshelev) [inline common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala]: Why trace? I would use info

File: common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    config.getString("messaging.mode") match {
+      case "nats" => startNats(system, config)
+      case "kinesis" =>
+        log.trace("Starting Kinesis")
+        KinesisClient.runWorker(system, config, system.log)
+        KinesisClient.createPublisher(system, config, system.log)
+      case "test" =>
+        Xor.right((msg:DeviceSeenMessage) => system.eventStream.publish(msg))
+      case _ =>
```
Comment (koshelev) [inline common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala]: Defaulting to NATS does not makes sense to me. Why not just fail?

File: common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        system.log.warning("Unknown messaging type detected, defaulting to NATS")
+        startNats(system, config)
+    }
+  }
+
+  def parseMsg(json: String): Xor[io.circe.Error, DeviceSeenMessage] = {
```
Comment (koshelev) [inline common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala]: Parsing messages should not be a responsibility of this object

File: common-messaging/src/main/scala/org/genivi/sota/messaging/kinesis/RecordProcessor.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  override def initialize(shardId: String): Unit = {
+    log.trace("Initializing kinesis worker")
+  }
+
+  /* We don't use checkpointer as the messages handled for now can be sent twice */
+  override def processRecords(records: util.List[Record], checkpointer: IRecordProcessorCheckpointer): Unit = {
+    for(record <- records) {
+      jawn.parseByteBuffer(record.getData) match {
```
Comment (koshelev) [inline common-messaging/src/main/scala/org/genivi/sota/messaging/kinesis/RecordProcessor.scala]: please use for-complrehension

File: core/src/main/scala/org/genivi/sota/core/DeviceUpdatesResource.scala [hunk 7]
```diff
@@ -62,9 +65,15 @@ class DeviceUpdatesResource(db: Database,
 
   protected lazy val updateService = new UpdateService(DefaultUpdateNotifier, deviceRegistry)
 
+  protected val publishFn = MessageBusManager.start(system, ConfigFactory.load()) match {
```
Comment (koshelev) [inline core/src/main/scala/org/genivi/sota/core/DeviceUpdatesResource.scala]: `getOrElse` is more readable here

Comment (calumats) [inline core/src/main/scala/org/genivi/sota/core/DeviceUpdatesResource.scala]: It's not clear to me how I would use getOrElse without throwing away the ConfigException that start() can return.

File: common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala (hunk 4 above)
Comment (calumats) [inline common-messaging/src/main/scala/org/genivi/sota/messaging/MessageBusManager.scala]: Fair point. I'll just fail.

**Comment by koshelev** [comment]
Pull Request (4/26)
  - [x] There must exist a JIRA ticket that describes the motivation for the code being submitted
  - [ ] The JIRA ticket must be referenced in the pull request message
  - [x] Commits must be squashed
  - [x] Review feedback must have been addressed

Branch (3/26)

- [ ] Branches must be correctly named:
- [x] Branch name must have a 'feat/' or 'bugfix/' or 'refactor/' prefix
- [ ] Branch name must include the name of the JIRA ticket

Code style / comments (4/26)

- [ ] Code must be formatted in accordance with style guide for the project
- [x] Any 'TODO' or 'FIXME' comments must reference a JIRA ticket
- [x] Code changes must have matching comment changes where required
- [ ] There must be no commented-out or dormant code

Tests (6/26)

- [ ] Any new functionality must be covered by tests
- [ ] Automated tests must exist, as far as possible, to test the acceptance criteria
- [ ] As far as it makes sense, tests should reference the JIRA ticket
- [ ] Where the JIRA ticket corresponds to top-level user story, there must be an end-to-end integration test
- [ ] The end-to-end integration test must reference the associated JIRA ticket in the code
- [x] CI tests must have run and passed

Requirements (4/26)

- [x] The functionality implemented must match the description in the JIRA ticket
- [ ] If the implementation is richer than the JIRA ticket description, additional detail should be added to the ticket
- [x] Any security controls / mitigations identified in the JIRA ticket must have been implemented
- [ ] Any acceptance criteria identified in the JIRA ticket must be met by the code

Documentation (5/26)

- [ ] Documentation must be up to date, specifically
- [ ] Any change to system architecture must be reflected in architecture diagrams
- [ ] Any changes to bootstrap, test or deployment requirements must be documented
- [ ] Any changes to API end-points or contracts must be documented
- [ ] All public functions, classes and files are appropriately commented

**Comment by koshelev** [comment]
Review Score: 9/26

**Pull request merged**